src/twisted/internet/protocol.py

```diff
@@ -30,9 +30,7 @@ class Factory:
     """
 
     # Put a subclass of Protocol here:
-    protocol = (
-        None
-    )  # type: Optional[Union[Type['Protocol'], Type['AbstractDatagramProtocol']]]  # noqa
+    protocol = None  # type: Optional[Type["Protocol"]]
```

Comment:
I previously added `AbstractDatagramProtocol` to this annotation, because I ran into mypy problems elsewhere without it.  But if you can get rid of that here, that would be definitely simplify things.

Reply:
Yeah, [this ignore comment](https://github.com/twisted/twisted/pull/1430/files#diff-ad9e865d47a56bc625d757669be5663cL1331) is an example of fallout from undoing this.

I prefer that, though as ignore comments are a thing to look back at and fix, whereas I wouldn't find this annotation in a search for ways to improve the type hints; I'd have to run across it and wonder why it's so complex and then look into it and then decide I care…

Reply:
This is why it's always a good idea to write up a bug and include a link to it next to any type:ignore's it affects :)

Reply:
@glyph can you clarify what you would prefer the type of the protocol variable to be here?

1.   Leave it as `Optional[Union[Type['Protocol'], Type['AbstractDatagramProtocol']]] `
2.  change it to `type: Optional[Type["Protocol"]]`, and add `type:ignore[assignment]` comments in the tree as per: https://github.com/twisted/twisted/pull/1430/files#diff-ad9e865d47a56bc625d757669be5663cL1331

Reply:
All of these types seem to be wrong, based on the intent of the attribute.  Hopefully `Optional[Callable[[], IProtocol]]` won't provoke any errors?  (It is depressing that `Factory` itself doesn't document this attribute when it's clearly so widely used…)

Reply:
In my tree, I tried changing to:

```
diff --git a/src/twisted/internet/protocol.py b/src/twisted/internet/protocol.py
index 058644eef..1433105d6 100644
--- a/src/twisted/internet/protocol.py
+++ b/src/twisted/internet/protocol.py
@@ -11,12 +11,12 @@ Twisted.  The Protocol class contains some introductory material.
 
 
 import random
-from typing import Optional, Type, Union
+from typing import Optional, Type, Union, Callable
 from zope.interface import implementer
 
 from twisted.python import log, failure, components
 from twisted.internet import interfaces, error, defer
-from twisted.internet.interfaces import ITransport
+from twisted.internet.interfaces import ITransport, IProtocol
 from twisted.logger import _loggerFor
 
 
@@ -32,7 +32,7 @@ class Factory:
     # Put a subclass of Protocol here:
     protocol = (
         None
-    )  # type: Optional[Union[Type['Protocol'], Type['AbstractDatagramProtocol']]]  # noqa
+    )  # type: Optional[Callable[[], IProtocol]]
 
     numPorts = 0
     noisy = True
diff --git a/src/twisted/test/test_tcp.py b/src/twisted/test/test_tcp.py
index 4ae90494b..04bf70ffb 100644
--- a/src/twisted/test/test_tcp.py
+++ b/src/twisted/test/test_tcp.py
@@ -11,7 +11,7 @@ import random
 import errno
 import hamcrest
 from functools import wraps
-from typing import Optional, Type, Union
+from typing import Optional, Type, Union, Callable
 from unittest import skipIf
 
 from zope.interface import implementer
@@ -22,7 +22,7 @@ from twisted.python.log import msg, err
 from twisted.internet import protocol, reactor, defer, interfaces
 from twisted.internet import error
 from twisted.internet.address import IPv4Address
-from twisted.internet.interfaces import IHalfCloseableProtocol, IPullProducer
+from twisted.internet.interfaces import IHalfCloseableProtocol, IPullProducer, IProtocol
 from twisted.protocols import policies
 from twisted.python.runtime import platform
 from twisted.test.proto_helpers import AccumulatingProtocol
@@ -122,7 +122,7 @@ class MyProtocolFactoryMixin:
     protocolConnectionLost = None
     protocol = (
         None
-    )  # type: Optional[Union[Type[protocol.Protocol],Type[protocol.AbstractDatagramProtocol]]]  # noqa
+    )  # type: Optional[Callable[[], IProtocol]]
     called = 0
 
     def __init__(self):
```


and got these mypy errors:

```
src/twisted/words/protocols/irc.py:2826:16: error: Incompatible types in assignment (expression has type "Type[DccSendProtocol]", base class "Factory" defined the type as "Optional[Callable[[], IProtocol]]")  [assignment]
        protocol = DccSendProtocol
                   ^
src/twisted/words/protocols/irc.py:2938:16: error: Incompatible types in assignment (expression has type "Type[DccChat]", base class "Factory" defined the type as "Optional[Callable[[], IProtocol]]")  [assignment]
        protocol = DccChat
                   ^
src/twisted/internet/endpoints.py:199:16: error: Incompatible types in assignment (expression has type "Type[_WrappingProtocol]", base class "Factory" defined the type as "Optional[Callable[[], IProtocol]]")  [assignment]
        protocol = _WrappingProtocol
                   ^
src/twisted/words/protocols/jabber/xmlstream.py:688:16: error: Incompatible types in assignment (expression has type "Type[XmlStream]", base class "Factory" defined the type as "Optional[Callable[[], IProtocol]]")  [assignment]
        protocol = XmlStream
                   ^
src/twisted/names/server.py:64:16: error: Incompatible types in assignment (expression has type "Type[DNSProtocol]", base class "Factory" defined the type as "Optional[Callable[[], IProtocol]]")  [assignment]
        protocol = dns.DNSProtocol
                   ^
src/twisted/web/proxy.py:91:16: error: Incompatible types in assignment (expression has type "Type[ProxyClient]", base class "Factory" defined the type as "Optional[Callable[[], IProtocol]]")  [assignment]
        protocol = ProxyClient
                   ^
src/twisted/test/test_amp.py:90:1: error: Cannot determine consistent method resolution order (MRO) for "SimpleSymmetricProtocol"  [misc]
    class SimpleSymmetricProtocol(amp.AMP):
    ^
src/twisted/test/test_amp.py:110:1: error: Cannot determine consistent method resolution order (MRO) for "TransportPeer"  [misc]
    class TransportPeer(amp.Argument):
    ^
src/twisted/test/test_amp.py:131:9: error: List item 3 has incompatible type "Tuple[bytes, TransportPeer]"; expected "Tuple[bytes, Argument]"  [list-item]
            (b"from", TransportPeer(optional=True)),
            ^
src/twisted/test/test_amp.py:131:19: error: Unexpected keyword argument "optional" for "TransportPeer"  [call-arg]
            (b"from", TransportPeer(optional=True)),
                      ^
src/twisted/test/test_amp.py:158:9: error: List item 3 has incompatible type "Tuple[bytes, TransportPeer]"; expected "Tuple[bytes, Argument]"  [list-item]
            (b"from", TransportPeer(optional=True)),
            ^
src/twisted/test/test_amp.py:158:19: error: Unexpected keyword argument "optional" for "TransportPeer"  [call-arg]
            (b"from", TransportPeer(optional=True)),
                      ^
src/twisted/test/test_amp.py:246:1: error: Cannot determine consistent method resolution order (MRO) for "FactoryNotifier"  [misc]
    class FactoryNotifier(amp.AMP):
    ^
src/twisted/test/test_amp.py:411:1: error: Cannot determine consistent method resolution order (MRO) for "LiteralAmp"  [misc]
    class LiteralAmp(amp.AMP):
    ^
src/twisted/test/test_amp.py:730:1: error: Cannot determine consistent method resolution order (MRO) for "TestLocator"  [misc]
    class TestLocator(amp.CommandLocator):
    ^
src/twisted/test/test_amp.py:766:1: error: Cannot determine consistent method resolution order (MRO) for "OverrideLocatorAMP"  [misc]
    class OverrideLocatorAMP(amp.AMP):
    ^
src/twisted/test/test_amp.py:2166:1: error: Cannot determine consistent method resolution order (MRO) for "NormalCommandProtocol"  [misc]
    class NormalCommandProtocol(amp.AMP):
    ^
src/twisted/test/test_amp.py:2178:1: error: Cannot determine consistent method resolution order (MRO) for "InheritedCommandProtocol"  [misc]
    class InheritedCommandProtocol(amp.AMP):
    ^
src/twisted/test/test_amp.py:2191:1: error: Cannot determine consistent method resolution order (MRO) for "AddedCommandProtocol"  [misc]
    class AddedCommandProtocol(amp.AMP):
    ^
src/twisted/test/test_amp.py:2431:1: error: Cannot determine consistent method resolution order (MRO) for "ProtocolIncludingArgument"  [misc]
    class ProtocolIncludingArgument(amp.Argument):
    ^
src/twisted/test/test_amp.py:2464:18: error: List item 0 has incompatible type "Tuple[bytes, ProtocolIncludingArgument]"; expected "Tuple[bytes, Argument]"  [list-item]
        arguments = [(b"weird", ProtocolIncludingArgument())]
                     ^
src/twisted/test/test_amp.py:2465:17: error: List item 0 has incompatible type "Tuple[bytes, ProtocolIncludingArgument]"; expected "Tuple[bytes, Argument]"  [list-item]
        response = [(b"weird", ProtocolIncludingArgument())]
                    ^
src/twisted/test/test_amp.py:2505:1: error: Cannot determine consistent method resolution order (MRO) for "NoNetworkProtocol"  [misc]
    class NoNetworkProtocol(amp.AMP):
    ^
src/twisted/pair/test/test_tuntap.py:1331:24: error: Incompatible types in assignment (expression has type "Type[IPRecordingProtocol]", variable has type "Optional[Callable[[], IProtocol]]")  [assignment]
        factory.protocol = IPRecordingProtocol
                           ^
src/twisted/pair/test/test_tuntap.py:1353:24: error: Incompatible types in assignment (expression has type "Type[EthernetRecordingProtocol]", variable has type "Optional[Callable[[], IProtocol]]")  [assignment]
        factory.protocol = EthernetRecordingProtocol
                           ^
Found 25 errors in 7 files (checked 840 source files)
ERROR: InvocationError for command /Users/crodrigues/Twisted/build/mypy/bin/mypy --cache-dir=/Users/crodrigues/Twisted/build/mypy_cache --pretty src (exited with code 1)
```

Reply:
Also `AbstractDatagramProtocol` does not implement `connectionLost()` which is required by the `IProtocol` interface.  That's why I used the type `Optional[Union[Type['Protocol'], Type['AbstractDatagramProtocol']]]`, because there are a bunch of places that assume you can set `protocol` to an `AbstractDatagramProtocol`

Reply:
- `DccSendFactory` appears to be … totally unused. Undocumented everywhere. `DccSendProtocol`, similarly so. Not clear what is trying to be expressed here.  The type error is real; this only works because `DccSendProtocol` redefines the expected type of `protocol` and thereby breaks LSP. Also it looks like it might be a security nightmare if anybody actually tried to use this, so maybe just… deprecate & delete?

Reply:
- `DccChatFactory` is similarly defining a completely new interface. It shouldn't be subclassing `ClientFactory`, it's not compatible.  It should be *implementing* an `IClientFactory` maybe.  Although it's similarly unused, at least it's tested.  Again fixing the types on the IRC side would make sense here.

Reply:
- `_WrappingFactory` once again demonstrates the need for an `IClientFactory` which doesn't have this protocol attribute at all.  It's also changing the contract and should change the type of the declared attribute.

Reply:
- `XmlStream` just has a bogus type declaration; delete it and it should work

Reply:
- `DNSServerFactory`'s `buildProtocol` once again changes the contract; this is a real type error

Reply:
- `ProxyClientFactory` once again changes the contract in its `buildProtocol`. Same as other instances.

Reply:
- `test_tuntap.py`'s usages are goofy nonsense.  In real life these datagram protocols are constructed and passed directly to `TuntapPort`.  (See for example `docs/pair/examples/pairudp.py`).  The real implementation does not use `Factory` and here `Factory` with `.protocol` set to `lambda: y` having `buildProtocol` called is merely being used as a shortcut for `(lambda: y)()`.  Eliminate the `factory` from the tests entirely, just instantiate the `Protocol` or make a method that returns it for each teste.

Reply:
So yeah, `AbstractDatagramProtocol` is a complete red herring here, the type should remain `Optional[Callable[[], IProtocol]]`.

Reply:
@glyph: I'm assuming that we're going to need to go through all of the `type: ignore` items after we've typed most of Twisted, so I feel like it's a bit premature to file tickets for all of them, because some of those get fixed by better typing elsewhere in the code, and others as actual bugs, etc.

My feeling on this one is to leave it as it is now, and fix the weirdness in `AbstractDatagramProtocol` in a different PR.

Reply:
Sorry, let me back-pedal a bit.  I see why `Optional[Callable[[], IProtocol]]` seems more correct, and we should start with more correct.  Let me start there.

Reply:
One problem with `IProtocol`:

```console
src/twisted/internet/protocol.py:132:9: error: "IProtocol" has no attribute "factory"  [attr-defined]
            p.factory = self
```

It seems like `Factory`  and `Protocol` are rather coupled, so `IProtocol` seems inappropriate here.

Reply:
If you can, I would suggest reverting all protocol-related changes from this PR.
I am OK with all other changes in this PR, so if we could get those in that would be good.

Then in a separate PR, tackle all of @glyph's feedback related to protocol.

Reply:
OK. Check out 0a10890 for fixes to the above, which includes comments referring to new tickets about wrongness noted by @glyph.

9b6fc8a deprecates the unused stuff in IRC.  I included `DccChatFactory` because it's not used in Twisted and I also couldn't find users on GitHub.

Reply:
My suggestion is that https://github.com/twisted/twisted/commit/0a10890695d0236f9171883f8dfaa934e97f7221 and https://github.com/twisted/twisted/commit/9b6fc8a5a53a92d4e29335b1898ebe48d2e98546 be pulled out of this PR into separate PR's, since this is beginning to have too many changes beyond just adding type hints.

Deprecated could requires a `.removal` newsfragment and tests as per the [Twisted compatibilty policy](https://twistedmatrix.com/documents/current/core/development/policy/compatibility-policy.html)

Reply:
0a10890 is required it's just ignores plus comments referring to the tickets, as @glyph requested.

I can pull 9b6fc8a if you like, but it's a pretty small change. Lemme know and I'll either do that or add the `.removal` thing.

Reply:
https://github.com/twisted/twisted/commit/9b6fc8a5a53a92d4e29335b1898ebe48d2e98546 needs to be a separate PR with a proper newsfragment and tests as per the [Twisted compatibilty policy](https://twistedmatrix.com/documents/current/core/development/policy/compatibility-policy.html)

Reply:
OK I'll pull it

Reply:
Done. I'll file a PR for it after this is merged.